### PR TITLE
feat/servicebus: issue3254 enhance router module and servicebus

### PR DESCRIPTION
**Pull request by ttlv**
**What this PR does / why we need it**:
https://github.com/kubeedge/kubeedge/issues/3254
**Which issue(s) this PR fixes**:
https://github.com/kubeedge/kubeedge/issues/3254
```

### CODE REVIEW

File: edge/pkg/servicebus/dao/targeturls.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2020 The KubeEdge Authors.
```
Comment (fisherxu) [inline edge/pkg/servicebus/dao/targeturls.go:2]:
```suggestion
Copyright 2021 The KubeEdge Authors.
```

File: edge/pkg/servicebus/dao/targeturls_test.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2020 The KubeEdge Authors.
```
Comment (fisherxu) [inline edge/pkg/servicebus/dao/targeturls_test.go:2]:
```suggestion
Copyright 2021 The KubeEdge Authors.
```

**Review by fisherxu — COMMENTED** [review]
Please make the CI passed first. :)

File: edge/pkg/servicebus/dao/targeturls_test.go (hunk 2 above)
Comment (ttlv) [reply edge/pkg/servicebus/dao/targeturls_test.go]: fixd :)

**Comment by ttlv** [comment]
> Please make the CI passed first. :)

CI has passed now :)

File: cloud/pkg/router/provider/servicebus/servicebus.go [hunk 3]
```diff
@@ -21,38 +24,97 @@ type servicebusFactory struct{}
...
+	message := data.(*model.Message)
+	res := make(map[string]interface{})
+	v, ok := message.Content.(string)
+	if !ok {
+		return nil, errors.New("message content invalid convert to string")
+	}
+	res["data"] = []byte(v)
+	resp, err := target.GoToTarget(res, nil)
+	if err != nil {
+		klog.Errorf("message is send to target fail. msgID: %s, target: %s, err:%v", message.GetID(), target.Name(), err)
```
Comment (gy95) [inline cloud/pkg/router/provider/servicebus/servicebus.go:92]: send message to target failed?

File: edge/pkg/servicebus/dao/targeturls_test.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+	cases := []struct {
+		// name is name of the testcase
+		name string
+		// queryTableReturn is the return of mock interface ormerMock's QueryTable function
+		queryTableReturn orm.QuerySeter
+		// result of the function
+		Result bool
+	}{{
+		// not enpty
```
Comment (gy95) [inline edge/pkg/servicebus/dao/targeturls_test.go:154]: not empty

File: edge/pkg/servicebus/dao/targeturls_test.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+		name:             "count > 0",
+		queryTableReturn: querySeterMock,
+		Result:           false,
+	},
+		{
+			// enpty
```
Comment (gy95) [inline edge/pkg/servicebus/dao/targeturls_test.go:160]: empty

File: edge/pkg/servicebus/dao/targeturls.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+type TargetUrls struct {
+	URL string `orm:"column(url);type(text);pk"`
+}
+
+// InsertUrls insert target_urls
+func InsertUrls(url string) error {
+	_, err := dbm.DBAccess.Raw("INSERT OR REPLACE INTO target_urls (url) VALUES (?)", url).Exec()
+	klog.V(4).Infof("INSERT result %v", err)
```
[the remark this answers is not in the document]
Comment (ttlv) [reply edge/pkg/servicebus/dao/targeturls.go:35]: just log the result of insert like other module such as eventbus.

File: cloud/pkg/router/provider/servicebus/servicebus.go (hunk 3 above)
Comment (ttlv) [reply cloud/pkg/router/provider/servicebus/servicebus.go]: fixed :)

File: edge/pkg/servicebus/dao/targeturls_test.go (hunk 4 above)
Comment (ttlv) [reply edge/pkg/servicebus/dao/targeturls_test.go]: fixed :)

File: edge/pkg/servicebus/dao/targeturls_test.go (hunk 5 above)
Comment (ttlv) [reply edge/pkg/servicebus/dao/targeturls_test.go]: fixed :)

File: edge/pkg/servicebus/servicebus.go [hunk 7]
```diff
@@ -88,72 +121,154 @@ func processMessage(msg *beehiveModel.Message) {
...
+	}
+}
+
+func server(stopChan <-chan struct{}) {
+	var (
+		timeout time.Duration
+		err     error
+	)
+	if timeout, err = time.ParseDuration(fmt.Sprintf("%vs", servicebusConfig.Config.Timeout)); err != nil {
+		klog.Errorf("can't format timeout and the default value will be setted")
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:209]: setted -> set

File: edge/pkg/servicebus/servicebus.go [hunk 8]
```diff
@@ -88,72 +121,154 @@ func processMessage(msg *beehiveModel.Message) {
...
+			serverResponse.Code = http.StatusBadRequest
+			serverResponse.Msg = fmt.Sprintf("url %v is not allowed and please make a rule for this url in the cloud", serverRequest.TargetURL)
+			w.Write(marshalResult(serverResponse))
+			return
+		}
+		msg := beehiveModel.NewMessage("").BuildRouter("servicebus", modules.UserGroup,
+			serverRequest.TargetURL, beehiveModel.UploadOperation).FillBody(byteData)
+		responseMessage, err := beehiveContext.SendSync(modules.EdgeHubModuleName, *msg, timeout)
+		resp, ok := responseMessage.GetContent().(string)
+		if err != nil {
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:259]: the two lines need to change the order

File: edge/pkg/servicebus/servicebus.go [hunk 9]
```diff
@@ -88,72 +121,154 @@ func processMessage(msg *beehiveModel.Message) {
...
+		Handler: h,
+	}
+	go func() {
+		<-stopChan
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		if err := s.Shutdown(ctx); err != nil {
+			klog.Errorf("Server shutdown failed: %s", err)
+		}
+		inited = 0
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:225]: whether do we need to use `atomic.CompareAndSwapInt32` method here, and I doubt if we need to use atomic methods to process the int variable `inited`

File: edge/pkg/servicebus/servicebus.go [hunk 10]
```diff
@@ -88,72 +121,154 @@ func processMessage(msg *beehiveModel.Message) {
...
-	}
-
-	var httpRequest commonType.HTTPRequest
-	if err := json.Unmarshal(content, &httpRequest); err != nil {
-		m := "error to parse http request"
-		code := http.StatusBadRequest
-		klog.Errorf(m, err)
-		if response, err := buildErrorResponse(msg.GetID(), m, code); err == nil {
-			beehiveContext.SendToGroup(modules.HubGroup, response)
+		content, err := json.Marshal(msg.GetContent())
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:144]: can use `GetContentData` instead.

File: edge/pkg/servicebus/servicebus.go [hunk 11]
```diff
@@ -88,72 +121,154 @@ func processMessage(msg *beehiveModel.Message) {
...
+				beehiveContext.SendToGroup(modules.HubGroup, response)
+			}
+			return
 		}
-		return
+
+		response := commonType.HTTPResponse{Header: resp.Header, StatusCode: resp.StatusCode, Body: resBody}
+		responseMsg := beehiveModel.NewMessage(msg.GetID())
+		responseMsg.Content = response
+		responseMsg.SetRoute("servicebus", modules.UserGroup).SetResourceOperation("", beehiveModel.UploadOperation)
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:198]: beehiveModel.NewMessage(msg.GetID()).FillBody(response).SetRoute(.........)

Comment (gy95) [reply edge/pkg/servicebus/servicebus.go:198]: there are three string literal "servicebus" in this file, and can be replaced with definitions `modules.ServiceBusModuleName`

File: edge/pkg/servicebus/servicebus.go [hunk 12]
```diff
@@ -1,48 +1,77 @@
...
+	timeout int
+}
+
+type ServerRequest struct {
+	Method    string `json:"method"`
+	TargetURL string `json:"targetURL"`
+	Payload   string `json:"payload"`
+}
+
+type ServerResponse struct {
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:52]: sry, I don't understand two structures, I only find where we process two structures. 
can you explain where we construct these structures,

File: edge/pkg/servicebus/servicebus.go [hunk 13]
```diff
@@ -88,72 +121,154 @@ func processMessage(msg *beehiveModel.Message) {
...
-	content, err := json.Marshal(msg.GetContent())
-	if err != nil {
-		klog.Errorf("marshall message content failed %v", err)
-		m := "error to marshal request msg content"
-		code := http.StatusBadRequest
-		if response, err := buildErrorResponse(msg.GetID(), m, code); err == nil {
-			beehiveContext.SendToGroup(modules.HubGroup, response)
+	case "stop":
+		dao.DeleteUrlsByKey(resource)
+		c <- struct{}{}
```
Comment (gy95) [inline edge/pkg/servicebus/servicebus.go:132]: and we can only process one URL one time?

File: edge/pkg/servicebus/servicebus.go (hunk 12 above)
Comment (gy95) [reply edge/pkg/servicebus/servicebus.go:52]: i know

Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: we construct these two structures in buildBasicHandler func to handler request from other application or send response to other application.servicebus become a universal http server to handler any call from any other application

File: edge/pkg/servicebus/servicebus.go (hunk 13 above)
Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: yeap,that's my fault. We should judge if the table target_urls is empty，if true we will send stop to the channel for stopping the servicebus server

File: edge/pkg/servicebus/servicebus.go (hunk 11 above)
Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: okay it had been fixed :)

File: edge/pkg/servicebus/servicebus.go (hunk 10 above)
Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: 👍

File: edge/pkg/servicebus/servicebus.go (hunk 9 above)
Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: yeap you are right.

File: edge/pkg/servicebus/servicebus.go (hunk 8 above)
Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: fixed :)

File: edge/pkg/servicebus/servicebus.go (hunk 7 above)
Comment (ttlv) [reply edge/pkg/servicebus/servicebus.go]: okay

**Comment by gy95** [comment]
why merge this commit `do not maintain nodestatus in localcache`

**Comment by ttlv** [comment]
> why merge this commit `do not maintain nodestatus in localcache`

because my error rebase operation and i will open a new pr and close this pr

**Pull request closed without merging**